metrics/certificaterequest_collector.go

```diff
@@ -0,0 +1,249 @@
+	expiryTime := 0.0
+
+	if cr.Status.Certificate != nil {
+		notAfter, err := getCertNotAfterTime(cr.Status.Certificate)
+		if err != nil {
+			return
+		}
+		expiryTime = float64(notAfter.Unix())
+	}
+
+	metric := prometheus.MustNewConstMetric(
+		cc.certificateRequestExpirationTimestampSeconds,
+		prometheus.GaugeValue,
+		expiryTime,
+		cr.Name,
+		cr.Namespace,
+		cr.Spec.IssuerRef.Name,
+		cr.Spec.IssuerRef.Kind,
+		cr.Spec.IssuerRef.Group,
+	)
+
+	ch <- metric
+}
+
+// updateCertificateRequestRenewalTime updates the renewal time metric for the given certificate request.
+// The renewal time is the time at which the volume should be renewed.
+// Note: there might be multiple certificate requests for a volume depending on the MaxRequestsPerVolume value,
+// but only the latest one will be stored in the nextIssuanceTimeMap.
+func (cc *CertificateRequestCollector) updateCertificateRequestRenewalTime(cr *cmapi.CertificateRequest, nextIssuanceTimeMap map[string]time.Time, ch chan<- prometheus.Metric) {
+	renewalTime := 0.0
+
+	if len(cr.Labels) != 0 {
+		if nextIssuanceTime, ok := nextIssuanceTimeMap[cr.Labels[internalapi.VolumeIDHashLabelKey]]; ok {
+			renewalTime = float64(nextIssuanceTime.Unix())
+		}
+	}
+
+	metric := prometheus.MustNewConstMetric(
+		cc.certificateRequestRenewalTimestampSeconds,
+		prometheus.GaugeValue,
+		renewalTime,
+		cr.Name,
+		cr.Namespace,
+		cr.Spec.IssuerRef.Name,
+		cr.Spec.IssuerRef.Kind,
+		cr.Spec.IssuerRef.Group,
+	)
+
+	ch <- metric
+}
+
+// getCertNotAfterTime returns the NotAfter time of the issued certificate.
+// It expects the certificate to be encoded in PEM format.
+func getCertNotAfterTime(certBytes []byte) (time.Time, error) {
+	block, _ := pem.Decode(certBytes)
+	if block == nil {
+		return time.Time{}, fmt.Errorf("invalid PEM data: could not decode certificate")
+	}
+	crt, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("parsing issued certificate: %w", err)
+	}
+
+	return crt.NotAfter, nil
+}
+
+// getNextIssuanceTimeMapFromMetadata returns a map of volume ID hashes to the next issuance time.
+// The map is keyed by the volume ID hash.
+// The next issuance time is the time at which the volume should be renewed.
+func (cc *CertificateRequestCollector) getNextIssuanceTimeMapFromMetadata() (map[string]time.Time, error) {
+	volumeIDs, err := cc.metadataReader.ListVolumes()
+	if err != nil {
+		return nil, fmt.Errorf("listing volumes: %w", err)
+	}
+
+	nextIssuanceTimeMap := make(map[string]time.Time, len(volumeIDs))
+	for _, id := range volumeIDs {
+		volumeMetadata, err := cc.metadataReader.ReadMetadata(id)
+		if err != nil {
+			return nil, err
+		}
+		if volumeMetadata.NextIssuanceTime != nil {
+			nextIssuanceTimeMap[internalapiutil.HashIdentifier(id)] = *volumeMetadata.NextIssuanceTime
+		}
+	}
+	return nextIssuanceTimeMap, nil
+}
+
+func (cc *CertificateRequestCollector) updateManagedVolumeCount(count int, ch chan<- prometheus.Metric) {
```

Comment:
@7ing I feel this should be `Gauge` instead of a `Counter` since we calculate `len(nextIssuanceTimeMap)` each time. I would love to know why it should be a Counter.